package/src/utilities/add-virtual-imports.ts

```diff
@@ -7,32 +7,79 @@ const resolveVirtualModuleId = <T extends string>(id: T): `\0${T}` => {
 	return `\0${id}`;
 };
 
-const createVirtualModule = (
-	name: string,
-	imports: Record<string, string>,
-): Plugin => {
+type VirtualImport = {
+	id: string;
+	content: string;
+	context?: "server" | "client" | undefined;
+};
+
+type Imports = Record<string, string> | Array<VirtualImport>;
+
+const createVirtualModule = (name: string, _imports: Imports): Plugin => {
 	const pluginName = `vite-plugin-${name}`;
 
+	// We normalize the imports into an array
+	const imports: Array<VirtualImport> = Array.isArray(_imports)
+		? _imports
+		: Object.entries(_imports).map(([id, content]) => ({
+				id,
+				content,
+				context: undefined,
+		  }));
+
+	// We check for virtual imports with overlapping contexts, eg. several imports
+	// with the same id and context server
+	const duplicatedImports: Record<string, Array<string>> = {};
```

Comment:
This could have some cheeky tests against it, but I'm not too bothered since everything is still quite changey

Reply:
yay and i don't know how to write yet so they'll have to wait 😅